### PR TITLE
Increase fidelity of rternalize signature

### PR DESCRIPTION
**Pull request by krassowski**
Exploring a solution for #905.

Not sure about preferred testing strategy (in addition to the two tests already added), but one approach could be:

```python
def dput(obj):
    connection = base.textConnection(NULL, open='w')
    base.dput(obj, file=connection)
    return base.gsub(' ', '', base.paste(
        base.textConnectionValue(connection),
        collapse=''
    ))

def f(a, /, b, c='c', *args, d='d', e, **kwargs):
    return dput(base.list(
        a=a,
        b=b,
        c=c,
        d=d,
        e=e,
        args=base.c(*args),
        kwargs=base.c(**kwargs)
    ))

f_r = rternalize_with_signature(f)

# for various signatures
assert f('a', 'b', e='e')[0] == f_r('a', 'b', e='e')[0]
f_r('A', 'B', e='E', kwarg1='KW1', kwarg2='KW2')[0] == 'list(a="A",b="B",c="c",d="d",e="E",args=NULL,kwargs=c(kwarg1="KW1",kwarg2="KW2"))'
```

### CODE REVIEW

**Comment by lgautier** [comment]
It is a great start. I have added candidate additional improvements. @krassowski - Can you have a look? If I missed something we can revert.

**Comment by lgautier** [comment]
> It is a great start. I have added candidate additional improvements. @krassowski - Can you have a look? If I missed something we can revert.

Bump. After a bit of struggle with merges and keeping the linter happy the change seems to have a clean bill. Can you have a look?

File: rpy2/rinterface.py [hunk 1]
```diff
@@ -1097,26 +1098,149 @@ def _post_initr_setup() -> None:
...
+    The following Python function can be "rternalized":
+
+    .. code-block:: python
+       def foo(x, *args, **kwargs):
+           pass
+
+    It is visible to R with the signature
+
+    .. code-block:: r
+       function (x, ..., z)
```
Comment (krassowski) [inline rpy2/rinterface.py:1153]: What is `z`?

File: rpy2/rinterface.py [hunk 2]
```diff
@@ -1097,26 +1098,149 @@ def _post_initr_setup() -> None:
...
+    is a valid Python declaration. However, it cannot be "rternalized"
+    because the R ellipsis can only appear at most once in the signature
+    of a given function. Trying to apply the decorator `rternalize` would
+    raise an exception.
+
+    The following Python function can be "rternalized":
+
+    .. code-block:: python
+       def foo(x, *args, **kwargs):
+           pass
```
Comment (krassowski) [inline rpy2/rinterface.py:1148]:
```suggestion
    The following Python function (equivalent for most purposes) can be "rternalized":

    .. code-block:: python
       def foo(x, *args, **kwargs):
           y = kwargs.get('y', 2)
           pass
```

**Comment by krassowski** [comment]
From a cursory look it seems ok, but I could not check out locally (some issue with my environment getting a seg fault on `_str_to_charsxp` even on main branch, so unrelated to this PR; unfortunately I don't have more time to investigate now).

File: rpy2/rinterface.py (hunk 1 above)
Comment (lgautier) [reply rpy2/rinterface.py:1153]: Good question. This is a mistake. The R signature will be `function (x, ...)`.

File: rpy2/rinterface.py (hunk 2 above)
Comment (lgautier) [reply rpy2/rinterface.py:1148]: I am happy to include the suggestion. I don't understand it though. Is there text missing? Why add `kwarg.get('y', 2)?

**Comment by lgautier** [comment]
> From a cursory look it seems ok, but I could not check out locally (some issue with my environment getting a seg fault on `_str_to_charsxp` even on main branch, so unrelated to this PR; unfortunately I don't have more time to investigate now)

Thanks. Let's wait a bit then. I am potentially breaking what you'd like to have. It is better we take the time needed for you to look at it.

**Comment by lgautier** [comment]
I have tried a bit more. It seems to keep what the first version of the PR did, plus a little more. I am merging it. We can address possible misses later.

**Comment by lgautier** [comment]
Of course issues appeared as soon as I merged.

The example with `dplyr` (in the doc) was no longer working, but I had assumed that it was because of a minor change in `dplyr`.

@krassowski - I have added an optional named argument "signature" to `rternalize`.
It works this way:
```python
# No mapping of signature. That's the way it was before.
@rternalize    # Can also be @rternalize(signature=False)
def foo(x):
    pass 

# Mapping of signatures.
@rternalize(signature=True):
def foo(x):
    pass
```

**Pull request merged**